File: src/main/java/net/dumbdogdiner/dogcore/commands/EconomyCommands.java
package net.dumbdogdiner.dogcore.commands;

import java.util.concurrent.CompletionStage;
import java.util.function.Function;
import net.dumbdogdiner.dogcore.Permissions;
import net.dumbdogdiner.dogcore.database.User;
import net.dumbdogdiner.dogcore.messages.Messages;
import net.kyori.adventure.text.Component;
import org.bukkit.OfflinePlayer;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.NotNull;
import revxrsal.commands.annotation.Command;
import revxrsal.commands.annotation.Default;
import revxrsal.commands.annotation.Optional;
import revxrsal.commands.annotation.Range;
import revxrsal.commands.annotation.Subcommand;
import revxrsal.commands.bukkit.annotation.CommandPermission;

public final class EconomyCommands {
    private EconomyCommands() { }

    /**
     * The /balance command.
     * @param sender The sender of the command.
     * @param player The player to check the balance of.
     */
    @Command({"balance", "bal"})
    @CommandPermission(Permissions.ECO)
    public static void balance(final CommandSender sender, @Optional final OfflinePlayer player) {
        OfflinePlayer playerToCheck;
        if (player == null) {
            if (sender instanceof Player p) {
                playerToCheck = p;
            } else {
                sender.sendMessage(Messages.get("error.playerNeeded"));
                return;
            }
        } else {
            playerToCheck = player;
        }

        User.lookupCommand(playerToCheck, sender, user -> user.getBalance().thenAccept(balance -> {
            if (sender.equals(playerToCheck)) {
                sender.sendMessage(Messages.get("commands.balance.query", Component.text(balance)));
            } else {
                user.formattedName().thenAccept(name ->
                    sender.sendMessage(Messages.get("commands.balance.query.other", name, Component.text(balance))));
            }
        }));
    }

    /**
     * The /balancetop command.
     * @param sender The sender of the command.
     * @param page The page to check.
     */
    @Command({"balancetop", "baltop"})
    @CommandPermission(Permissions.ECO)
    public static void balanceTop(final CommandSender sender, @Default("1") @Range(min = 1.0) final int page) {
        User.top(page).thenAccept(entries -> {
            for (var entry : entries) {
                sender.sendMessage(
                    Messages.get("commands.balancetop.entry", entry.name(), Component.text(entry.amount())));
            }
        });
    }

    /**
     * The /pay command.
     * @param sender The sender of the command.
     * @param player The player to send money to.
     * @param amount The amount of money to send.
     */
    @Command("pay")
    @CommandPermission(Permissions.ECO)
    public static void pay(
        final Player sender,
        final OfflinePlayer player,
        @Range(min = 1.0) final long amount
    ) {
        User.lookupCommand(sender, sender,
            from -> User.lookupCommand(player, sender, to -> from.pay(to, amount).thenAccept(success -> {
                if (success) {
                    to.formattedName().thenAccept(name ->
                        sender.sendMessage(Messages.get("commands.pay.success", Component.text(amount), name)));
                } else {
                    sender.sendMessage(Messages.get("error.failedTransaction"));
                }
            })));
    }

    private static void economyHelper(
        @NotNull final CommandSender sender,
        @NotNull final OfflinePlayer player,
        @NotNull final Function<@NotNull User, @NotNull CompletionStage<@NotNull Boolean>> action
    ) {
        User.lookupCommand(player, sender, user -> action.apply(user).thenAccept(success -> {
            if (success) {
                sender.sendMessage(Messages.get("commands.economy.success"));
            } else {
                sender.sendMessage(Messages.get("error.failedTransaction"));
            }
        }));
    }

    @Command({"economy", "eco"})
    @Subcommand("give")
    @CommandPermission(Permissions.ECO_ADMIN)
    public static void economyGive(
        final CommandSender sender,
        final OfflinePlayer player,
        @Range(min = 1.0) final long amount
    ) {
        economyHelper(sender, player, user -> user.give(amount));
    }

    @Command({"economy", "eco"})
    @Subcommand("take")
    @CommandPermission(Permissions.ECO_ADMIN)
    public static void economyTake(
        final CommandSender sender,
        final OfflinePlayer player,
        @Range(min = 1.0) final long amount
    ) {
        economyHelper(sender, player, user -> user.give(-amount));
    }

    @Command({"economy", "eco"})
    @Subcommand("set")
    @CommandPermission(Permissions.ECO_ADMIN)
    public static void economySet(
        final CommandSender sender,
        final OfflinePlayer player,
        final long amount
    ) {
        economyHelper(sender, player, user -> user.setBalance(amount).thenApply(v -> true));
    }
}

File: src/main/kotlin/net/dumbdogdiner/dogcore/commands/EconomyCommands.kt
package net.dumbdogdiner.dogcore.commands

import net.dumbdogdiner.dogcore.Permissions
import net.dumbdogdiner.dogcore.db.DbPlayer
import net.kyori.adventure.text.Component
import org.bukkit.OfflinePlayer
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player
import revxrsal.commands.annotation.Command
import revxrsal.commands.annotation.Optional
import revxrsal.commands.annotation.Range
import revxrsal.commands.annotation.Subcommand
import revxrsal.commands.bukkit.annotation.CommandPermission

object EconomyCommands {
    @Command("balance", "bal")
    @CommandPermission(Permissions.ECO)
    fun balance(
        sender: CommandSender,
        @Optional
        player: OfflinePlayer?
    ) {
        val p = player
            ?: sender as? Player
            ?: run {
                sender.sendMessage(Component.text("Player argument needed in this context."))
                return
            }
        val dbPlayer = DbPlayer.lookup(p)
        if (dbPlayer == null) {
            sender.sendMessage(Component.text("Player has no records in the server."))
            return
        }
        val balance = dbPlayer.balance
        if (p == sender) {
            sender.sendMessage(Component.text("You have $balance bean(s)."))
        } else {
            sender.sendMessage(Component.text("${dbPlayer.username} has $balance bean(s)."))
        }
    }

    @Command("balancetop", "baltop")
    @CommandPermission(Permissions.ECO)
    fun balanceTop(
        sender: CommandSender,
        @Optional
        @Range(min = 1.0)
        page: Int?
    ) {
        for ((username, balance) in DbPlayer.top(page ?: 1)) {
            sender.sendMessage(Component.text("$username -> $balance"))
        }
    }

    @Command("pay")
    @CommandPermission(Permissions.ECO)
    fun pay(
        sender: Player,
        player: OfflinePlayer,
        @Range(min = 1.0)
        amount: Long
    ) {
        val s = DbPlayer.lookup(sender) ?: return
        val p = DbPlayer.lookup(player)
            ?: run {
                sender.sendMessage(Component.text("Cannot pay a player who has never joined."))
                return
            }
        if (s.pay(p, amount)) {
            sender.sendMessage(Component.text("You paid $amount bean(s) to ${p.username}."))
        } else {
            sender.sendMessage(Component.text("Could not complete the transaction."))
        }
    }

    private fun economyHelper(sender: CommandSender, player: OfflinePlayer, action: (DbPlayer) -> Boolean) {
        val p = DbPlayer.lookup(player)
            ?: run {
                sender.sendMessage(Component.text("Player argument needed in this context."))
                return
            }
        if (action(p)) {
            sender.sendMessage(Component.text("The balance was updated"))
        } else {
            sender.sendMessage(Component.text("The balance would overflow, so it was not updated."))
        }
    }

    @Command("economy", "eco")
    @Subcommand("give")
    @CommandPermission(Permissions.ECO_ADMIN)
    fun economyGive(
        sender: CommandSender,
        player: OfflinePlayer,
        @Range(min = 1.0)
        amount: Long
    ) {
        economyHelper(sender, player) { it.give(amount) }
    }

    @Command("economy", "eco")
    @Subcommand("take")
    @CommandPermission(Permissions.ECO_ADMIN)
    fun economyTake(
        sender: CommandSender,
        player: OfflinePlayer,
        @Range(min = 1.0)
        amount: Long
    ) {
        economyHelper(sender, player) { it.give(-amount) }
    }

    @Command("economy", "eco")
    @Subcommand("set")
    @CommandPermission(Permissions.ECO_ADMIN)
    fun economySet(sender: CommandSender, player: OfflinePlayer, amount: Long) {
        economyHelper(sender, player) {
            it.balance = amount
            true
        }
    }
}
